src/cpp.cc

```diff
@@ -2122,16 +2167,49 @@ void Printer::Cpp::multi_print_cyk(
   stream << indent() << *t << " t_" << track << "_n = t_" << real_track
     << "_seq.size();" << endl << endl;
 
+  bool checkpoint = ast->checkpoint && ast->checkpoint->cyk;
+
   // SMJ 2023-02-18: I've added the x: xxx loops comments, but
   // am not 100% sure if they are correct in multitrack contexts
   if (!inner.empty()) {
-    stream << indent() << "for (" << *t << " " << js << " = 0; " << js << " < "
+    if (checkpoint) {
+      /*
+         in checkpointing mode, the loop indices (e.g. t_0_i and t_0_j
+         in single-track mode; names can differ in multi-track mode)
+         are members of the out class instead of local loop variables
+         so they can be archived/loaded;
+         if they are loaded, the outer loop variable (e.g. t_0_j) starts
+         from whatever value was loaded from the checkpoint
+      */
+      stream << indent() << "for (; " << js << " < "
       << ns << "; " << "++" << js << ") {" << endl;
+    } else {
+      stream << indent() << "for (" << *t << " " << js << " = 0; "
+             << js << " < " << ns << "; " << "++" << js << ") {"
+             << endl;
+    }
     inc_indent();
+
     stream << indent() << "// A: quadratic loops" << endl;
-    stream << indent() << "for (" << *t << " " << is << " = " << js << " + 1; "
-      << is << " > 1; " << is << "--) {" << endl;
-    inc_indent();
+    if (checkpoint) {
+      /*
+         in checkpointing mode, the inner loop variable (e.g. t_0_i)
+         is either regularly set to e.g. t_0_j + 1 or to whatever value
+         was loaded from the checkpoint archive;
+         however, this loaded checkpoint value will only be the start value
+         for t_0_i in the first inner loop pass;
+         afterwards it will regularly be set to t_0_j + 1;
+      */
+      stream << indent() << "for (" << is << " = ("
+             << is << " == 1 ? " << js << " + 1 : " << is << "); "
+             << is << " > 1; " << is << "--) {" << endl;
```

Comment:
will this carry over to multitrack grammars, e.g. https://github.com/jlab/ADP_collection/blob/master/alignments.gap? There you can have #tracks * 2 nested loops

Reply:
It should, but I'm not sure. I will test that.